File: nim/test/test_roundtrip.nim
## test_roundtrip.nim - Roundtrip benchmark test for MPC Designation Converter (Nim)

import ../src/mpc_designation
import os, strutils, times

proc main() =
  if paramCount() < 1:
    echo "Usage: test_roundtrip <csv_file>"
    quit(1)

  let csvFile = paramStr(1)
  if not fileExists(csvFile):
    echo "Error: File not found: ", csvFile
    quit(1)

  # Load all data first
  var unpackedData: seq[string]
  var packedData: seq[string]
  var first = true

  for line in lines(csvFile):
    if first:
      first = false
      continue  # Skip header

    let parts = line.split(',')
    if parts.len < 2:
      continue

    unpackedData.add(parts[0])
    packedData.add(parts[1])

  let total = unpackedData.len
  echo "Processing ", total, " entries"
  echo ""

  # Phase 1: Pack (unpacked -> packed)
  echo "=== Phase 1: Pack ==="
  var packPassed = 0
  var packFailed = 0

  let packStart = cpuTime()
  for i in 0..<total:
    let result = convertSimple(unpackedData[i])
    if result == packedData[i]:
      inc packPassed
    else:
      inc packFailed
  let packEnd = cpuTime()
  let packElapsed = (packEnd - packStart) * 1000
  let packRate = if packElapsed > 0: float(total) * 1000.0 / packElapsed else: 0.0

  echo "Passed: ", packPassed
  echo "Failed: ", packFailed
  echo "Time: ", int(packElapsed), "ms (", formatFloat(packRate, ffDecimal, 1), " entries/sec)"
  echo ""

  # Phase 2: Unpack (packed -> unpacked)
  echo "=== Phase 2: Unpack ==="
  var unpackPassed = 0
  var unpackFailed = 0

  let unpackStart = cpuTime()
  for i in 0..<total:
    let result = convertSimple(packedData[i])
    if result == unpackedData[i]:
      inc unpackPassed
    else:
      inc unpackFailed
  let unpackEnd = cpuTime()
  let unpackElapsed = (unpackEnd - unpackStart) * 1000
  let unpackRate = if unpackElapsed > 0: float(total) * 1000.0 / unpackElapsed else: 0.0

  echo "Passed: ", unpackPassed
  echo "Failed: ", unpackFailed
  echo "Time: ", int(unpackElapsed), "ms (", formatFloat(unpackRate, ffDecimal, 1), " entries/sec)"
  echo ""

  # Round-trip verification: pack(unpack(packed)) == packed
  echo "=== Round-trip Verification ==="
  var rtPassed = 0
  var rtFailed = 0

  for i in 0..<total:
    let unpacked = convertSimple(packedData[i])
    let repacked = convertSimple(unpacked)
    if repacked == packedData[i]:
      inc rtPassed
    else:
      inc rtFailed

  if rtFailed == 0:
    echo "Packed RT: PASS (", rtPassed, "/", total, ")"
  else:
    echo "Packed RT: FAIL (", rtFailed, " failures)"

when isMainModule:
  main()

File: php/test/test_roundtrip.php
#!/usr/bin/env php
<?php
/**
 * Test mpc_designation.php roundtrip conversion.
 *
 * Tests that pack(unpack(x)) == x for packed designations,
 * and unpack(pack(x)) == x for unpacked designations.
 *
 * Usage: php test_roundtrip.php <csv_file> [max_errors]
 */

declare(strict_types=1);

require_once __DIR__ . '/../src/MPCDesignation.php';

use MPC\MPCDesignation;
use MPC\MPCDesignationException;

function runRoundtripTests(string $csvFile, int $maxErrors = 100): bool {
    $total = 0;
    $passed = 0;
    $failed = 0;
    $errors = [];

    $startTime = microtime(true);

    $fp = fopen($csvFile, 'r');
    if ($fp === false) {
        fwrite(STDERR, "Error: Cannot open file: $csvFile\n");
        return false;
    }

    // Skip header line
    fgets($fp);

    while (($line = fgets($fp)) !== false) {
        $line = trim($line);
        if ($line === '') {
            continue;
        }

        // Skip header if encountered
        if (str_starts_with($line, 'unpacked')) {
            continue;
        }

        $total++;
        $parts = explode(',', $line);
        $unpacked = $parts[0];
        $packed = $parts[1];

        // Test packed -> unpacked -> packed roundtrip
        try {
            $toUnpacked = MPCDesignation::unpack($packed);
            $backToPacked = MPCDesignation::pack($toUnpacked);

            if ($backToPacked !== $packed) {
                $failed++;
                if (count($errors) < $maxErrors) {
                    $errors[] = ["packed roundtrip", $packed, $toUnpacked, $backToPacked, $packed];
                }
                continue;
            }
        } catch (MPCDesignationException $e) {
            $failed++;
            if (count($errors) < $maxErrors) {
                $errors[] = ["packed roundtrip", $packed, "ERROR", $e->getMessage(), $packed];
            }
            continue;
        }

        // Test unpacked -> packed -> unpacked roundtrip
        // Note: old-style designations like "A908 CJ" convert to modern "1908 CJ"
        try {
            $toPacked = MPCDesignation::pack($unpacked);
            $backToUnpacked = MPCDesignation::unpack($toPacked);

            // For old-style designations, the unpacked form changes
            // Check if it's an old-style designation
            $isOldStyle = preg_match('/^[AB]\d{3} [A-Z][A-Z]$/', $unpacked);

            if (!$isOldStyle && $backToUnpacked !== $unpacked) {
                $failed++;
                if (count($errors) < $maxErrors) {
                    $errors[] = ["unpacked roundtrip", $unpacked, $toPacked, $backToUnpacked, $unpacked];
                }
                continue;
            }

            // For old-style, verify the modern form roundtrips
            if ($isOldStyle) {
                $repackedModern = MPCDesignation::pack($backToUnpacked);
                if ($repackedModern !== $toPacked) {
                    $failed++;
                    if (count($errors) < $maxErrors) {
                        $errors[] = ["old-style modern roundtrip", $unpacked, $backToUnpacked, $repackedModern, $toPacked];
                    }
                    continue;
                }
            }
        } catch (MPCDesignationException $e) {
            $failed++;
            if (count($errors) < $maxErrors) {
                $errors[] = ["unpacked roundtrip", $unpacked, "ERROR", $e->getMessage(), $unpacked];
            }
            continue;
        }

        $passed++;

        // Progress indicator every 100,000 entries
        if ($total % 100000 === 0) {
            echo "Processed $total entries...\n";
        }
    }

    fclose($fp);

    $elapsed = microtime(true) - $startTime;

    echo "\n";
    echo "=== Roundtrip Test Results ===\n";
    echo "Total:  $total\n";
    echo "Passed: $passed\n";
    echo "Failed: $failed\n";
    $elapsedMs = $elapsed * 1000;
    $rate = $total / $elapsed;
    echo sprintf("Time:   %.0fms (%.1f entries/sec)\n", $elapsedMs, $rate);
    echo "\n";

    if ($failed > 0) {
        echo sprintf("=== First %d failures ===\n", count($errors));
        echo sprintf("%-20s %-20s %-15s %-20s %-15s\n", 'Test', 'Input', 'Step1', 'Step2', 'Expected');
        echo str_repeat('-', 90) . "\n";
        foreach ($errors as [$test, $input, $step1, $step2, $expected]) {
            echo sprintf("%-20s %-20s %-15s %-20s %-15s\n", $test, $input, $step1, $step2, $expected);
        }
    }

    return $failed === 0;
}

function main(array $argv): int {
    if (count($argv) < 2) {
        fwrite(STDERR, "Usage: php test_roundtrip.php <csv_file> [max_errors]\n");
        return 1;
    }

    $csvFile = $argv[1];
    $maxErrors = isset($argv[2]) ? (int)$argv[2] : 100;

    $success = runRoundtripTests($csvFile, $maxErrors);
    return $success ? 0 : 1;
}

exit(main($argv));
